server/src/middleware/rpc/layers.rs

```diff
@@ -172,3 +179,60 @@ where
 		rp
 	}
 }
+
+/// Similar to [`tower::util::Either`] but
+/// adjusted to satisfy the traits bound [`RpcServiceT].
+#[pin_project(project = EitherProj)]
+#[derive(Clone, Debug)]
+pub enum Either<A, B> {
+	/// One type of backing [`RpcServiceT`].
+	A(#[pin] A),
+	/// The other type of backing [`RpcServiceT`].
+	B(#[pin] B),
+}
+
+impl<A, B, T, AE, BE> Future for Either<A, B>
+where
+	A: Future<Output = Result<T, AE>>,
+	AE: Into<Box<dyn StdError + Send + Sync>>,
+	B: Future<Output = Result<T, BE>>,
+	BE: Into<Box<dyn StdError + Send + Sync>>,
+{
+	type Output = Result<T, Box<dyn StdError + Send + Sync>>;
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+		match self.project() {
+			EitherProj::A(fut) => Poll::Ready(Ok(ready!(fut.poll(cx)).map_err(Into::into)?)),
+			EitherProj::B(fut) => Poll::Ready(Ok(ready!(fut.poll(cx)).map_err(Into::into)?)),
+		}
+	}
+}
+
+impl<S, A, B> Layer<S> for Either<A, B>
+where
+	A: Layer<S>,
+	B: Layer<S>,
+{
+	type Service = Either<A::Service, B::Service>;
+
+	fn layer(&self, inner: S) -> Self::Service {
+		match self {
+			Either::A(layer) => Either::A(layer.layer(inner)),
+			Either::B(layer) => Either::B(layer.layer(inner)),
+		}
+	}
+}
+
+#[async_trait::async_trait]
+impl<'a, A, B> RpcServiceT<'a> for Either<A, B>
+where
+	A: RpcServiceT<'a> + Send + Sync,
+	B: RpcServiceT<'a> + Send + Sync,
+{
+	async fn call(&self, request: Request<'a>, transport: TransportProtocol) -> MethodResponse {
```

Comment:
Dq unrelated to this PR : do we want to expose in the future extra information together with the `TransportPortocol` to the user? 🤔 Generally I'm thinking towards a `CallContext` that has `fn transport_protocol(&self) -> &TransportProtocol`